### PR TITLE
Convert the build system to meson, probe headers rather than guessing based on libc/kernel 

### PR DESCRIPTION
**Pull request by ArsenArsen**
While porting Gentoo to mlibc, I ran into quite a bit of trouble with pax-utils. Primary one of these was the reliance on macros like `__GLIBC__` to determine which system headers to include, which doesn't work for us for obvious reasons. In the process of trying to fix that, I discovered that the build system was a bit of a mess, and ended up rewriting the entire thing.

I also fixed the indentation in generated manpages and some warnings.

Note that I am not entirely sure that this behaviour matches the previous perfectly, so it might require a careful review.

Thanks!

### CODE REVIEW

**Comment by thesamesam** [comment]
@SoapGentoo

**Comment by grobian** [comment]
For whatever my opinion is worth anything here, I'm very much against converting to meson.  I'd much rather see it autoconf'd instead if you need to probe for availability of headers, etc.  Things like meson put too much dependency pressure on a simple package like pax-utils.

**Comment by chewi** [comment]
I have no sway here either, but I'm of the opposite opinion. If you assume Python is a given (and with Gentoo, it has to be), then I believe Meson is lighter dependency-wise. I know this is used outside Gentoo, but Python at build-time is probably a given anywhere, and I suspect we've already burned this bridge.

**Comment by thesamesam** [comment]
> I have no sway here either, but I'm of the opposite opinion. If you assume Python is a given (and with Gentoo, it has to be), then I believe Meson is lighter dependency-wise. I know this is used outside Gentoo, but Python at build-time is probably a given anywhere, and I suspect we've already burned this bridge.

I was just writing this:
* meson only needs Python
* Portage obviously needs Python
* We only really need pax-utils for Portage anyway, and because of Portage, Python is already there

Plus there's muon (which is written in C, an implementation of meson) that can handle a fair amount of projects anyway.

Someone's done the work here (which I am very grateful for!) and the current build system doesn't really give us much flexibility - indeed, they hit real issues in trying to port to a new platform.

If someone wants to do the work for proper autotoolsification, I'll listen, but given someone's actually put up a Meson PR here, it feels a bit late for that. I don't really want to maintain two. If Arsen had rocked up with an autotools PR, I wouldn't have told him to go write it in Meson. But he offered Meson and here it is. 

I welcome opinions from anyone - but especially you guys as you're both involved in odd corners of Gentoo. So you _do_ have sway.

**Comment by grobian** [comment]
K, I guess the general concensus is in favour of meson.  So be it.  I'm happy to see there's a C-impl of meson, the dependencies of Python are growing and growing too.

**Comment by grobian** [comment]
> We only really need pax-utils for Portage anyway, and because of Portage, Python is already there

you do know scanelf/scanmacho are utilities that are far more useful than just for Portage alone, right? :)

**Comment by thesamesam** [comment]
Yep. So can you give an example of a situation where you want to use it, but Meson is prohibitive? Of course, it's only a build-time dependency anyway?

You didn't give an example originally? I don't know what constraints you're working within.

**Comment by SoapGentoo** [comment]
I really like this PR, and commend @ArsenArsen for the change. @grobian unless you give actual arguments why python is a problem here (and the fact that meson has no other dependencies except on the python interpreter) at build-time, I am strongly in favour of merging this and getting rid of the awful Makefiles.

**Comment by grobian** [comment]
pax-utils was written by the same folks that once used mainly embedded systems (they also wrote portage-utils) and on those systems, you typically don't have a bloated env like portage and python available, you just install packages in a root, or via a light-weight installer like qmerge (therefore also the original xpak-based binaries that were a simple uncompress/untar).

I don't know if that use-case is anywhere near remotely in use these days.  Given that we haven't heard in a long time from such scenarios, I guess it's safe to say newer versions can depend on more heavy-weight build-systems, like meson.

So, just to repeat: I raised my concerns, I have no real blocker arguments, so given there is a positive attitude from others, I am not going to stand in the way of making this happen.  Not sure I qualify as maintainer for pax-utils, though.

**Comment by ArsenArsen** [comment]
I can test and ensure Muon is able to interpret and build the project if you'd like me to (and add a CI workflow for it).

**Comment by SoapGentoo** [comment]
Yes, and your argument is wrong, because for such systems, meson would be in `BDEPEND` and used from a beefy system to compile pax-utils for some embedded system. The supposed "bloat" is on `CBUILD` and not `CHOST`, hence your argument is invalid.

**Comment by thesamesam** [comment]
> I can test and ensure Muon is able to interpret and build the project if you'd like me to (and add a CI workflow for it).

I think that might be a nice compromise. An interesting experiment if nothing else!

**Comment by grobian** [comment]
> I can test and ensure Muon is able to interpret and build the project if you'd like me to (and add a CI workflow for it).

You'd make me very happy, but I don't want to impose this is as a MUST.

**Comment by ArsenArsen** [comment]
Now this is interesting (from the muon build):

```
/home/runner/work/pax-utils/pax-utils/meson_options.txt:13:3: error expected ')', got 'identifier'
 13 |   description : 'Enable testing (not guaranteed to work)'
```

File: .github/workflows/build-test-ci.yml
Comment (thesamesam) [inline .github/workflows/build-test-ci.yml:58]: As an extra thought: if we want muon to work, we should probably test with samurai too (small ninja impl)

**Comment by ArsenArsen** [comment]
Alright, that should work now (tests still don't pass since the fuzzer target fails, but that was failing before also)

File: make-seccomp-filters.sh
Comment (xen0n) [inline make-seccomp-filters.sh:4]: This script could (and should) be wired into `meson.build` to ensure the `seccomp-bpf.h` is up-to-date? This way we could even remove and gitignore the `seccomp-bpf.h` altogether.

File: meson_options.txt
Comment (xen0n) [inline meson_options.txt:8]: Why not a feature? Although this is not controlling a dependency, it could be nice to be able to control this via `auto_features`.

File: make-seccomp-filters.sh
Comment (ArsenArsen) [reply make-seccomp-filters.sh:4]: Your two comments are related, so I'll bunch them together:

I opted not to change any behavior at all in this PR, so I submitted stuff matching the existing build system, hence a separate script for this (previously, it was a completely isolated depgraph in make)

The reason meson doesn't generate this script for us is because it'd add a dep on seccomp-bpf on the build machine (this is a problem outside of Linux, or even for particularly pedantic Linux users).

My personal choice for replacing this (as a less conservative change than what I was doing) would be to write the BPF by hand or to simply require seccomp-bpf at runtime.

Another option that was mentioned is dropping it altogether, as we don't really have any daemons here that run at high privileges anyway.

`use_seccomp` being a boolean rather than a feature is indeed since it doesn't dictate a dependency (hence auto makes little sense for it). I wanted to replace this with a check for -EINVAL at runtime but, again, decided to make everything as similar to before the change as humanly possible while still changing stuff.

File: .github/workflows/build-test-ci.yml
Comment (eli-schwartz) [inline .github/workflows/build-test-ci.yml]: Meson also uses `meson setup`.

The difference is that Meson has the IMHO surprising behavior of defaulting to an implicit `setup` if it doesn't recognize the second word as a subcommand. I really want to deprecate that behavior someday -- it means we don't have the freedom to create new subcommands because someone might use the same name as their build directory. We already had to choose "compile" instead of the original name of a "build" subcommand, because people often do `meson build` to specify a builddir at `./build/`.

In fact, you do that too :p but you have -D options in between. (To cater to muon, because muon is strict about POSIX utility syntax guidelines and requires the option-arguments to come before operands.)

Comment (ArsenArsen) [reply .github/workflows/build-test-ci.yml]: Hm, so common syntax would be `X setup -Da=b build`?

+1 on removing this behavior though ;)

Comment (eli-schwartz) [reply .github/workflows/build-test-ci.yml]: Yes, that's the common syntax.

**Review by xen0n — APPROVED** [review]
I've tried the meson build on my workstation and it worked fine. Thanks!

**Comment by ArsenArsen** [comment]
Great to hear! Thank you for testing

**Comment by thesamesam** [comment]
Please add signoff.

**Comment by ArsenArsen** [comment]
done

**Comment by vapier** [comment]
people seem to be confused.  pax-utils *was already autotooled*.  the bundled Makefile was purely intended for developers with modern/current systems.  everyone else, especially more esoteric build setups, were supposed to use standard configure+make.  in fact, this PR deleted the autogen.sh & configure.ac files which should have been a dead give away "it was already using autotools".  it has been using autotools since 2014 when i added it.

partially this is about a build system, and if muon didn't exist, this would be an obvious NAK.  but even with that, meson does not include the portability logic that we got completely for free with autotools & gnulib.  which is how pax-utils has been able to use more recent POSIX additions even when platforms we care about (e.g. macOS as @grobian noted) are stuck in the past.

**Comment by vapier** [comment]
ping ... if there's nothing in favor of keeping meson, i'll rollback to the portable autotools build that we had before

**Comment by thesamesam** [comment]
No issues have come up in the 2 years since this was merged. I didn't read your previous comment as saying "we should revert meson", just as commentary you wish you'd made before or similar? Nor have we had any issues on macOS since AFAIK...

I don't get the motivation here.

**Comment by SoapGentoo** [comment]
Sorry, NAK, Meson is here to stay.

**Comment by eli-schwartz** [comment]
> people seem to be confused.  pax-utils *was already autotooled*.  the bundled Makefile was purely intended for developers with modern/current systems.  everyone else, especially more esoteric build setups, were supposed to use standard configure+make.  in fact, this PR deleted the autogen.sh & configure.ac files which should have been a dead give away "it was already using autotools".  it has been using autotools since 2014 when i added it.
> 
> partially this is about a build system, and if muon didn't exist, this would be an obvious NAK.  but even with that, meson does not include the portability logic that we got completely for free with autotools & gnulib.  which is how pax-utils has been able to use more recent POSIX additions even when platforms we care about (e.g. macOS as @grobian noted) are stuck in the past.

I don't understand how you could possibly say that the autotools logic was already portable for free.

I also don't understand why you seem to think this PR claimed it wasn't using autotools.

The PR claimed that the existing build system was an un-portable mess. Using autotools doesn't make something portable -- using autotools correctly makes something portable.

It was not portable before. The PR submitter found it significantly easier to add portability using meson, rather than working with the existing autotools mess.

Reverting this PR and returning to the autotools build is a dereliction of your duty as a maintainer. However, reverting this PR and fixing up the autotools build to actually work portably could be argued as better than meson (that's a subjective decision, some people really like autotools and some people really like meson and ultimately maintainers gain consensus and decide, not external contributors). But crucially, it must actually be portable.

I leave discussion of "whether a correct meson or a correct autotools is preferred" to the people that actually have a voice in the matter, i.e. not Eli Schwartz.

**Comment by ArsenArsen** [comment]
the state of pax-utils I found it in was a homebrew makefile and a reportedly half-baked autoconf-based build system that portage wasn't even using.  I don't recall giving it much testing.  should you convert it to automake, I won't stop you, but I suspect I'll be alone in that.

(edit: note that ::gentoo used just the makefile)

**Comment by grobian** [comment]
Not sure, but would autotools next to meson be an option?  I did that for Portage's Prefix branch, and they coexist well there.

**Comment by eli-schwartz** [comment]
Sure, they can coexist fine.

Many projects even do this by default as a transition period. Provide both build systems for a short period while people test out the new one and make sure it fulfills all required use cases.

**Comment by grobian** [comment]
> Nor have we had any issues on macOS since AFAIK...

nit: apart from meson itself not working on macOS under certain situations -- now fixed.

FWIW, autotools & friends are more portable and friendly to older and diverse systems than new kids on the block, which often have to learn from the same mistakes autotools made in the past.  Add something like a language such as Python that requires a C11 compiler and you rapidly exclude on portability.  To give an idea, Perl might sound uncool, but it still compiles virtually anywhere without patching.  I'm not suggesting anyone to use Perl though, but autotools is (or was).  

So while pax-utils can compile fine on Mac OS X 10.5 code-wise, with meson as buildsystem it cannot.  You'd need Python 3.7, which in itself doesn't compile, likely will need a more recent compiler, thus also a linker, various libraries, such as libxml2, etc.  autotools is just a shell script using constructs that are old, thus supported by the system out of the box.  Comes at a cost for newer systems though.

It's all a matter of how available you want your software to be, and in what sense you feel it's ok to raise the bar.  The sentiment that is expressed here in this thread towards retaining support for non-standard things is not uncommon nowadays.  Given @vapier has been working on this code for a long time (nearly 2 decades), trying to keep it portable over the years, I don't see it as strange that he wants it to remain portable for as long as the code can easily be as such.

**Comment by vapier** [comment]
again, if you were using the makefile directly and complaining that it wasn't portable, you missed the point.  it wasn't meant to be portable, it was meant to be small & fast for people actually developing pax-utils.  if you wanted portable, use the autotools build.

people still continue to not understand what gnulib is or how it provided portability to the project, and have offered no alternative for pax-utils.  we leaned heavily on that in the past to use features that POSIX didn't offer, or only recently gained.  case in point: this PR decided that requiring `_GNU_SOURCE` was OK, and that everyone had to provide `asprintf`.  so pax-utils doesn't work on platforms that don't support those.

so people can either provide an alternative, or i'm rolling it back to autotools.

**Comment by eli-schwartz** [comment]
> Add something like a language such as Python that requires a C11 compiler and you rapidly exclude on portability.

> So while pax-utils can compile fine on Mac OS X 10.5 code-wise, with meson as buildsystem it cannot. You'd need Python 3.7, which in itself doesn't compile, likely will need a more recent compiler, thus also a linker, various libraries, such as libxml2, etc. 

Sorry, this is nonsense.

Python 3.11 requires c11, but python 3.7 -- which is what meson goes to ***great*** lengths to remain compatible with -- as well as 3.8, 3.9, and 3.10, use "C89 with several select C99 features" (six carefully listed features).

Python does not use libxml2. It does use expat, along with several other libraries, which it bundles a vendored copy of and will use that vendored library if a system copy doesn't exist.

I'd like to hear more about how Gentoo Prefix on macOS cannot compile python 3.7 (and especially, how does it install portage in that case).

**Comment by thesamesam** [comment]
> So while pax-utils can compile fine on Mac OS X 10.5 code-wise, with meson as buildsystem it cannot. You'd need Python 3.7, which in itself doesn't compile, likely will need a more recent compiler, thus also a linker, various libraries, such as libxml2, etc. autotools is just a shell script using constructs that are old, thus supported by the system out of the box. Comes at a cost for newer systems though.
> 

We have muon in CI and it works.

> It's all a matter of how available you want your software to be, and in what sense you feel it's ok to raise the bar. The sentiment that is expressed here in this thread towards retaining support for non-standard things is not uncommon nowadays. Given @vapier has been working on this code for a long time (nearly 2 decades), trying to keep it portable over the years, I don't see it as strange that he wants it to remain portable for as long as the code can easily be as such.

sure, but it would've been nice to have that comment before 2 years after it got merged, for a few reasons, but not least because other people use pax-utils and distros have migrated to meson already. It also wasn't clear to me at all (or seemingly anyone else who replied) that vapier's first comment warranted a reply. It just seemed like an observation.

> 
> people still continue to not understand what gnulib is or how it provided portability to the project, and have offered no alternative for pax-utils. we leaned heavily on that in the past to use features that POSIX didn't offer, or only recently gained. case in point: this PR decided that requiring `_GNU_SOURCE` was OK, and that everyone had to provide `asprintf`. so pax-utils doesn't work on platforms that don't support those.
> 
> so people can either provide an alternative, or i'm rolling it back to autotools.

What are you asking for here? You want an implementation of `asprintf` and to drop `_GNU_SOURCE`?  That's easy enough to do, we already do that in OpenRC.

Note that asprintf is in POSIX next too: https://austingroupbugs.net/view.php?id=1496.

**Comment by eli-schwartz** [comment]
> again, if you were using the makefile directly and complaining that it wasn't portable, you missed the point.  it wasn't meant to be portable, it was meant to be small & fast for people actually developing pax-utils.  if you wanted portable, use the autotools build.

Once again. The autotools build was NOT portable.

It is NOT portable to use autotools and then compile code that relies on

```
#if !defined(__FreeBSD__) && !defined(__NetBSD__) && !defined(__OpenBSD__)
```

If you want a portable autotools build then please actually make it portable. Thanks.

> people still continue to not understand what gnulib is or how it provided portability to the project, and have offered no alternative for pax-utils.  we leaned heavily on that in the past to use features that POSIX didn't offer, or only recently gained.  case in point: this PR decided that requiring `_GNU_SOURCE` was OK, and that everyone had to provide `asprintf`.  so pax-utils doesn't work on platforms that don't support those.
> 
> so people can either provide an alternative, or i'm rolling it back to autotools.

Gnulib doesn't provide portability to code that uses

```
#if !defined(__FreeBSD__) && !defined(__NetBSD__) && !defined(__OpenBSD__)
```

to determine what features are available on the system.

**Comment by grobian** [comment]
> Sorry, this is nonsense.

Sorry you feel this way.

> Python 3.11 requires c11, but python 3.7 -- which is what meson goes to _**great**_ lengths to remain compatible with -- as well as 3.8, 3.9, and 3.10, use "C89 with several select C99 features" (six carefully listed features).

Which is why I referenced Python 3.7, not 3.11.

> I'd like to hear more about how Gentoo Prefix on macOS cannot compile python 3.7 (and especially, how does it install portage in that case).

You can only do that with heavy patching and tweaking, which is beyond the conversation we're heaving here.

**Comment by eli-schwartz** [comment]
> > Sorry, this is nonsense.
> 
> Sorry you feel this way.
> 
> > Python 3.11 requires c11, but python 3.7 -- which is what meson goes to _**great**_ lengths to remain compatible with -- as well as 3.8, 3.9, and 3.10, use "C89 with several select C99 features" (six carefully listed features).
> 
> Which is why I referenced Python 3.7, not 3.11.

I do not understand why you're referencing c11 in the context of python 3.7.

If you believe I am incorrect for feeling this way then please explain to me:

Why is it *not* nonsense, to claim that python 3.7 requires a c11 compiler?

**Comment by thesamesam** [comment]
> It is NOT portable to use autotools and then compile code that relies on
> 
> ```
> #if !defined(__FreeBSD__) && !defined(__NetBSD__) && !defined(__OpenBSD__)
> ```

This is a solid chunk of why I was fine with it, as clearly it wasn't using gnulib in a bunch of places, and it was causing a real portability problem in porting to managarm.

**Comment by grobian** [comment]
> > It is NOT portable to use autotools and then compile code that relies on
> > ```
> > #if !defined(__FreeBSD__) && !defined(__NetBSD__) && !defined(__OpenBSD__)
> > ```
> 
> This is a solid chunk of why I was fine with it, as clearly it wasn't using gnulib in a bunch of places, and it was causing a real portability problem in porting to managarm.

Let there be no confusion here, that this indeed needed to be addressed, and it is good that this happened.

**Comment by grobian** [comment]
> > > Sorry, this is nonsense.
> > 
> > 
> > Sorry you feel this way.
> > > Python 3.11 requires c11, but python 3.7 -- which is what meson goes to _**great**_ lengths to remain compatible with -- as well as 3.8, 3.9, and 3.10, use "C89 with several select C99 features" (six carefully listed features).
> > 
> > 
> > Which is why I referenced Python 3.7, not 3.11.
> 
> I do not understand why you're referencing c11 in the context of python 3.7.
> 
> If you believe I am incorrect for feeling this way then please explain to me:


There is a confusement.  Python 3.11 requires C11.  Python 3.7 does not.  So C11 is not relevant there, nor in relation to meson.  Python 3.7 is the least required by meson, which is very nice.  However, if you only have Python 2.7, it means you'll have to update first, and I can guarantee you that Python 3.7 will not compile, or just crash horribly when it runs.  Python upstream doesn't care about that, they live in the realm of "only support shiny new things".

> 
> Why is it _not_ nonsense, to claim that python 3.7 requires a c11 compiler?

I never meant to claim that 3.7 requires C11.  In fact I thought I made a division so as to not suggest it did.

**Comment by grobian** [comment]
> > > > Sorry, this is nonsense.
> > > 
> > > 
> > > Sorry you feel this way.
> > > > Python 3.11 requires c11, but python 3.7 -- which is what meson goes to _**great**_ lengths to remain compatible with -- as well as 3.8, 3.9, and 3.10, use "C89 with several select C99 features" (six carefully listed features).
> > > 
> > > 
> > > Which is why I referenced Python 3.7, not 3.11.
> > 
> > 
> > I do not understand why you're referencing c11 in the context of python 3.7.
> > If you believe I am incorrect for feeling this way then please explain to me:
> 
There is a confusement. Python 3.11 requires C11. Python 3.7 does not. So C11 is not relevant there, nor in relation to meson. Python 3.7 is the least required by meson, which is very nice. However, if you only have Python 2.7, it means you'll have to update first, and I can guarantee you that Python 3.7 will not compile, or just crash horribly when it runs. Python upstream doesn't care about that, they live in the realm of "only support shiny new things".
> > Why is it _not_ nonsense, to claim that python 3.7 requires a c11 compiler?
I never meant to claim that 3.7 required C11.  In fact I thought I made a division to suggest this wasn't the case.

**Comment by ArsenArsen** [comment]
> again, if you were using the makefile directly and complaining that it wasn't portable, you missed the point. it wasn't meant to be portable, it was meant to be small & fast for people actually developing pax-utils. if you wanted portable, use the autotools build.

I was merely following suit WRT what Gentoo was using.  That path is the most tested path, and, in fact, the readme at the time stated:

> Just run `make`.  This should work on any recent POSIX compliant system.



> people still continue to not understand what gnulib is or how it provided portability to the project, and have offered no alternative for pax-utils.

I understand what gnulib is and how it provides portability (and have opted to use it for some WIP code that I'm writing for the purposes of Gentoo) but, indeed, have provided no alternative as the codebase massively relied on broken `defined(PLATFORM_MACRO)` tests (IIRC, I fixed some of those by replacing them with cc.check_header or w/e.. please keep those checks as AC_CHECK_HEADER/S if you go through with it), so I figured that the code was always intended to be GNUish-specific, at least IIRC.

> we leaned heavily on that in the past to use features that POSIX didn't offer, or only recently gained. case in point: this PR decided that requiring `_GNU_SOURCE` was OK, and that everyone had to provide `asprintf`. so pax-utils doesn't work on platforms that don't support those.

note that the makefile (which was the only build path documented in the readme and the one used in gentoo) did not use gnulib

> so people can either provide an alternative, or i'm rolling it back to autotools.

should you decide to do that, please keep the replacements for the `defined(PLATFORM_MACRO)` checks

**Comment by vapier** [comment]
@thesamesam i didn't say you did anything wrong.  you're fine.  don't worry about it.

> `#if !defined(__FreeBSD__) && !defined(__NetBSD__) && !defined(__OpenBSD__)`

that should be improved.  that didn't require meson to address.  however, those defines *still exist now*, and some of those will continue to exist forever.  because they're tied to how the ldso behaves, not whether the C library provides a specific API/header.

that's also a subset of visibility.  other headers provided by gnulib wouldn't have ifdef protection because we could assume they always exist.

> What are you asking for here? You want an implementation of asprintf and to drop _GNU_SOURCE? That's easy enough to do, we already do that in OpenRC.
> 
> Note that asprintf is in POSIX next too: https://austingroupbugs.net/view.php?id=1496.

that's great, but that's short sighted.  i don't want to write asprintf.  i don't want to write any of the other APIs that gnulib provided for free.  i want to use APIs from the next POSIX drop right away, not wait 2 decades.  glibc included asprintf in 2002.  POSIX will include it at some point ?  and then still have to refer to every platform we directly care about, and what other people care about, to provide it.  or we use gnulib and never think about it.

**Comment by vapier** [comment]
last call -- i'll prob look at it this weekend

**Comment by akhuettel** [comment]
> last call -- i'll prob look at it this weekend

NAK from toolchain. Meson stays.

**Comment by vapier** [comment]
the opinions of random people who have never contributed to pax-utils and lack any technical basis are irrelevant.  unless you have something useful to add to the discussion, please refrain from adding noise.

**Comment by thesamesam** [comment]
This has become very heated.

I really don't want to drop meson because it tends to be what people prefer now, and it's also churn for the downstream consumers of pax-utils. It also makes us look a bit silly to change our mind after 2 years and with no real-user complaint about it.

In the spirit of compromise, I can live with autotools done properly in parallel for fringe platforms if we must. In general, I get the point of gnulib, but it generally makes me uncomfortable because of how often we've had to go around patching ancient versions of it in packages for stuff like musl support -- sandbox even has quite an old copy of bundled gnulib without using the official `gnulib-tool`. Using gnulib is not maintenance-free and therefore "just using autotools" isn't without cost. We're against bundled libraries everywhere else, but for some reason, gnulib gets a pass?

**Comment by grobian** [comment]
Does `meson` have gnulib support?

Not using gnulib means not supporting "different" platforms, or writing support for it yourself -- which is silly if you can get it for free.

**Comment by ArsenArsen** [comment]
> Does `meson` have gnulib support?
> 
> Not using gnulib means not supporting "different" platforms, or writing support for it yourself -- which is silly if you can get it for free.

gnulib lacks meson support (note the inversion of responsibility), and likely will continue to do so, as meson lacks macros and functions.

re, gnulib bundling, I agree, and I think it'd be lovely if someone worked on a systemwide mode for gnulib (where it reuses a possibly patched systemwide copy of gnulib and prefers it to its own bundled one), for obvious reasons.  I've wanted to do this myself, but have so far been far too busy to focus any work on non-GNU systems

**Comment by eli-schwartz** [comment]
I would expect that gnulib works fine in meson as long as you use gnulib to vendor in the source files, then write your own copies of the equivalent build system tooling, i.e. port the m4 macros / Makefile rules to `meson.build` syntax.

Depends how much work it does at configure time vs. compile time, really, but the only difference one way or another is going to be time spent performing the vendoring vs. time someone else already did.

**Comment by grobian** [comment]
> re, gnulib bundling, I agree, and I think it'd be lovely if someone worked on a systemwide mode for gnulib (where it reuses a possibly patched systemwide copy of gnulib and prefers it to its own bundled one), for obvious reasons. I've wanted to do this myself, but have so far been far too busy to focus any work on non-GNU systems

Prefix team has tried this in a couple of ways and all of them failed one way or the other.  Gnulib is not made to be distributed as library.

**Comment by ArsenArsen** [comment]
> Prefix team has tried this in a couple of ways and all of them failed one way or the other. Gnulib is not made to be distributed as library.

yes, and Bruno Haible was open to fixing that last I asked him

**Comment by grobian** [comment]
The problem isn't so much gnulib team, it's that there are dozen (understatement) of packages which do not expect some systems to have certain functionality, or in a peculiar way, and by using gnulib's versions globally you break more packages than you actually fix.  So gnulib really is something you deliberately plug into your product, or not, which means you'll work around shortcomings.

**Comment by ArsenArsen** [comment]
> The problem isn't so much gnulib team, it's that there are dozen (understatement) of packages which do not expect some systems to have certain functionality, or in a peculiar way, and by using gnulib's versions globally you break more packages than you actually fix. So gnulib really is something you deliberately plug into your product, or not, which means you'll work around shortcomings.

by sharing the gnulib code between packages as well as its detections and replacements we don't magically break stuff that never tries to use gnulib, though?  it's not like we're installing libgnu.a as libc.so or something

but anyway, that's getting off-topic here.  both sam and I are quite interested in this and would like to get more folks on board (if only to be able to get more man-hours.. neither one of us has too much time), so I'd love to continue on this topic in -prefix or -toolchain

**Pull request closed without merging**
